src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/ReadOperation.java

```diff
@@ -77,6 +90,146 @@ default Uni<FindResponse> findDocument(
+                  query, Optional.ofNullable(stateRef.get()), pageSize);
+            })
+        // Read document while pagingState exists, limit for read is set at updateLimit +1
+        .whilst(resultSet -> extractPagingStateFromResultSet(resultSet) != null)
+        .onItem()
+        .transformToUniAndMerge(
+            resultSet -> {
+              Iterator<QueryOuterClass.Row> rowIterator =
+                  resultSet.getRowsList().stream().iterator();
+              int remaining = resultSet.getRowsCount();
+              documentCounter.addAndGet(remaining);
+              while (--remaining >= 0 && rowIterator.hasNext()) {
+                ReadDocument document = null;
+                QueryOuterClass.Row row = rowIterator.next();
+                List<JsonNode> sortValues = new ArrayList<>(numberOfOrderByColumn);
+                for (int sortColumnCount = 0;
+                    sortColumnCount < numberOfOrderByColumn;
+                    sortColumnCount++) {
+                  int columnCounter =
+                      SORTED_DATA_COLUMNS + ((sortColumnCount) * SORT_INDEX_COLUMNS_SIZE);
+                  QueryOuterClass.Value value = row.getValues(columnCounter);
+                  if (!value.hasNull()) {
+                    sortValues.add(nodeFactory.textNode(Values.string(value)));
+                    continue;
+                  }
+                  columnCounter++;
+                  value = row.getValues(columnCounter);
+                  if (!value.hasNull()) {
+                    sortValues.add(nodeFactory.numberNode(Values.decimal(value)));
+                    continue;
+                  }
+                  columnCounter++;
+                  value = row.getValues(columnCounter);
+                  if (!value.hasNull()) {
+                    sortValues.add(nodeFactory.booleanNode(Values.bool(value)));
+                    continue;
+                  }
+                  columnCounter++;
+                  value = row.getValues(columnCounter);
+                  if (!value.hasNull()) {
+                    sortValues.add(nodeFactory.nullNode());
+                    continue;
+                  }
+
+                  sortValues.add(nodeFactory.missingNode());
+                }
+                // Create ReadDocument with document id, grpc value for doc json and list of sort
+                // values
+                document =
+                    ReadDocument.from(
+                        getDocumentId(row.getValues(0)), // key
+                        row.getValues(1), // Deserialized value of doc_json
+                        sortValues);
+                sortedData.add(document);
+              }
+              return Uni.createFrom().item(true);
+            })
+        .collect()
+        .last()
+        .onItem()
+        .transform(
+            flag -> {
+              if (documentCounter.get() == errorLimit)
+                throw new JsonApiException(ErrorCode.DATASET_TOO_BIG);
```

Comment:
this is too late, you already have all the documents fetched..

Reply:
There is no other way to check it, unless we run count query. And we will fetch only (max allowed + 1) document

Reply:
Don't understand this? Yes you can count how many you already fetched in the first fetching loop.. Just have the state that counts saves next page state and the documents already fetched, if you are over the maximum, push exception down immediately..